### ISSUE TITLE
 Seeking feedback on Fān Qiáng (翻墙), the VPN protocol in the upcoming Awala VPN 

### ISSUE THREAD

**Issue by gnarea** [net4people/bbs#428]
Hey folks,

I'm gathering feedback on [the protocol](https://awala.app/en/vpn/tech-overview/) to power the upcoming [Awala VPN](https://awala.app/en/vpn/). I already built a working [prototype](https://github.com/relaycorp/fanqiang-poc) and tested it in China.

The Awala VPN will be a commercial service, but unlike all/most commercial VPNs:

- All the code will be publicly accessible, even the server-side stuff.
- I'm planning to make underlying obfuscation building blocks (e.g. the client library analogous to uTLS) open source so other apps can easily integrate it.
- In the future, tech-savvy individuals will be able to use our client with their own servers, skipping our infrastructure.

Looking forward to your thoughts!

Gus.

**Comment by iopq** [comment]
I didn't find it in the document, you are only planning to support TCP then? Because if you are trying to emulate browsing websites, it doesn't stop Chinese ISPs from disconnecting IPs that use too much UDP bandwidth

**Comment by gnarea** [comment]
The VPN client connects to the tunnel with WebSockets, so the GFW would see a TCP connection (with TLS). This connection encapsulates the actual IPv4/IPv6 packets, which could contain TCP/UDP/IMCP/etc payloads.

Please let me know if that answers your question!

**Comment by iopq** [comment]
If you tunnel UDP with websocket, you introduce TCP in UDP issues. For example, congestion control and retransmission of UDP when those applications are designed for low latency

**Comment by gnarea** [comment]
Agreed @iopq! That's a good point, and one I should've elaborated on when I said I wanted to migrate to MASQUE. I just updated the doc to [acknowledge the implications of using TCP for tunnelling](https://awala.app/en/vpn/tech-overview/#performance-considerations).

**Comment by gnarea** [comment]
Hi @smateoliu!

I'll be using Rust with an OpenSSL wrapper to avoid TLS client fingerprinting (potentially with [rquest](https://github.com/penumbra-x/rquest); TBC).

We're only using WebSockets initially, which only works on HTTP/1 (TCP), so [I'll be deferring the QUIC integration until we add support for MASQUE](https://awala.app/en/vpn/tech-overview/#masque-migration).

**Comment by gnarea** [comment]
Thanks for your feedback and questions @smateoliu! To answer your questions:

> The protocol design seems perfect and perfectly avoids client-side fingerprinting, but have you ever thought about how to avoid server-side port scanning (also known as active probing)? This means that the server-side program should not respond with anything (from TCP or UDP level) until a valid client attempts to connect and vaild data sent to server.

Indeed, that's why [we're using websites as tunnels](https://awala.app/en/vpn/tech-overview/#tunnel), under a hard-to-guess URL path (e.g. `https://example.com/s3cr3t`). We'll kick things off by operating those websites initially, but over time [we want to pay third-party website owners to run tunnels for us](https://awala.app/en/vpn/partners/tunnels/).

Our own tunnels won't have any other ports open, other than `80` (HTTP; redirects to HTTPS) and `443` (HTTPS); I'm unlikely to even run SSH because I prefer immutable servers. As for third-party tunnels, they may have other ports open, besides `80` and `443`, but that'd be irrespective of their role as an Awala VPN tunnel.

In other words, scanning the ports of the servers to which users connect, or even making HTTP(S) requests, wouldn't reveal the use of a tunnel.

> Have you considered how to ensure that clients in network-constrained environments can reliably connect to the API server? in network censorship contries. Governments can block primary domains via TCP resets, DNS pollution, QUIC packet dropping (filter by SNI). did you know?

Indeed. Once the user installs the app and connects to their allocated tunnels, their client will communicate with our API server via the tunnels too, to bypass censorship.

This means that the real challenge lies in the initial user onboarding. See: https://awala.app/en/vpn/tech-overview/#user-onboarding.

> Have you considered how to secure your application self-configuration to avoid other application read and analysis? this is really crtial point for windows. I don't know about other countries, but in China, Chinese antivirus software usually scans such software information and is used by the government.

I have, and unfortunately I don't think there's much we can do as mentioned in the threat [_VPN client detection via eavesdropping app_](https://awala.app/en/vpn/tech-overview/#vpn-client-detection-via-eavesdropping-app). I'd love to hear additional measures that we could consider.

> Are there plans to support international languages?

Absolutely! I'm planning to localise everything, including the apps, websites, and social media profiles, to as many regions as possible.

**Comment by iopq** [comment]
> Agreed @iopq! That's a good point, and one I should've elaborated on when I said I wanted to migrate to MASQUE. I just updated the doc to [acknowledge the implications of using TCP for tunnelling ](https://awala.app/en/vpn/tech-overview/#performance-considerations).

My concern with UDP connections is QoS/filtering UDP traffic, which is why I use https://github.com/dndx/phantun/ to pretend my UDP packets are TCP

**Comment by gnarea** [comment]
> My concern with UDP connections is QoS/filtering UDP traffic, which is why I use https://github.com/dndx/phantun/ to pretend my UDP packets are TCP

Sorry, but I don't understand how that applies here. The tunnels are powered by WebSockets, which means they use TCP. Are you concerned about UDP traffic throttled/blocked between [the gateway](https://awala.app/en/vpn/tech-overview/#gateway) and the target server? Or are you suggesting an alternative tunnelling protocol that doesn't use WebSockets?

**Comment by gnarea** [comment]
> Maybe I don't understand how MASQUE works well enough, isn't it over UDP?

Ah, got it! Yes, that's right. MASQUE operates over HTTP/3, which is built on top of QUIC, and QUIC uses UDP as its transport layer. We plan to support MASQUE in the future, but the plan is to go live with WebSockets only, for now.

I think you raised a good point about UDP traffic being potentially throttled/blocked, so [I added a note to the section on the MASQUE migration](https://awala.app/en/vpn/tech-overview/#masque-migration):

> (...) since MASQUE operates over UDP, it’s important to consider that some networks may throttle or block UDP traffic
